lib/ukrsi/rsi.c

```diff
@@ -0,0 +1,311 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/*
+ * Copyright (c) 2023, Xingjian Zhang <zhxj9823@qq.com>.
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ * 3. Neither the name of the copyright holder nor the names of its
+ *    contributors may be used to endorse or promote products derived from
+ *    this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <arm/smccc.h>
+#include <arch/rsi.h>
+#include <errno.h>
+#include <uk/essentials.h>
+#include <uk/arch/paging.h>
+#include <uk/plat/common/sections.h>
+extern __u64 mem_base, mem_sz;
+
+#ifdef CONFIG_PAGING
+#include <uk/arch/lcpu.h>
+#include <uk/plat/paging.h>
+#define RAM_BASE_ADDR 0x80000000
```

Comment:
This should not be hardcoded here. `plat/kvm/include/kvm-arm64/kvm-image.h` provides this value, you should take it from there. Also because this is conditional to `CONFIG_PAGING` the build fails if you disable paging.

Until `plat` is cleaned up you can do it as follows.

In `ukrsi/Makefile.uk` add:
```
LIBUKRSI_CINCLUDES-y += -I$(LIBKVMPLAT_BASE)/include
```

and in this file:
```
#include <kvm-arm64/image.h>
```